extensions/amp-a4a/0.1/test/test-amp-a4a.js

```diff
@@ -1141,6 +1142,89 @@ describe('amp-a4a', () => {
+      });
+
+      it('properly handles multiple providers that all fail', () => {
+        // Single provider with first key fails but second key passes validation
+        a4a.win.ampA4aValidationKeys = (() => {
+          const providers = [];
+          for (let i = 0; i < 10; i++) {
+            const serviceName = `test-service${i}`;
+            providers[i] = Promise.resolve({serviceName, keys: [
+              Promise.resolve({serviceName}),
+              Promise.resolve({serviceName}),
+            ]});
+          }
+          return providers;
+        })();
+        stubVerifySignature.returns(Promise.resolve(false));
+        return a4a.verifyCreativeSignature_('some_creative', 'some_sig')
+          .then(creative => {
+            expect(stubVerifySignature).to.be.callCount(20);
+            expect(creative).to.be.null;
+          });
+      });
+
+      it('properly handles multiple keys for one provider', () => {
+        // Single provider with first key fails but second key passes validation
+        const serviceName = 'test-service';
+        a4a.win.ampA4aValidationKeys = [
+          Promise.resolve({serviceName, keys: [
+            Promise.resolve({serviceName: '1'}),
+            Promise.resolve({serviceName: '2'}),
+          ]}),
+        ];
+        const creative = 'some_creative';
+        stubVerifySignature.onCall(0).returns(Promise.resolve(false));
+        stubVerifySignature.onCall(1).returns(Promise.resolve(true));
+        return a4a.verifyCreativeSignature_(creative, 'some_sig')
+          .then(verifiedCreative => {
+            expect(stubVerifySignature).to.be.calledTwice;
+            expect(verifiedCreative).to.equal(creative);
+          });
+      });
+
+      it('properly stops verification at first valid key', done => {
+        // Single provider where first key fails, second passes, and third
+        // never calls verifySignature.
+        const serviceName = 'test-service';
+        let keyInfoResolver;
+        a4a.win.ampA4aValidationKeys = [
+          Promise.resolve({serviceName, keys: [
+            Promise.resolve({}),
+            Promise.resolve({}),
+            new Promise(resolver => {
+              keyInfoResolver = resolver;
+            }),
+          ]}),
+        ];
+        const creative = 'some_creative';
+        const signature = 'some_signature';
+        stubVerifySignature.onCall(0).returns(Promise.resolve(false));
+        stubVerifySignature.onCall(1).returns(Promise.resolve(true));
+        a4a.verifyCreativeSignature_(creative, signature)
+          .then(verifiedCreative => {
+            expect(stubVerifySignature).to.be.calledTwice;
+            expect(verifiedCreative).to.equal(creative);
+            done();
+          });
+        // From testing have found that need to yield prior to calling last
+        // key info resolver to ensure previous keys have had a chance to
+        // execute.
+        setTimeout(() => {keyInfoResolver({}); }, 0);
```

Comment:
@taymonbeal is working on modifying key signature to ensure we know explicitly which provider and public key to use so this will be going away.  As a result, I would prefer not to invest too much into additional test coverage given priorities to work on other items.